File: premovechess.py
#This is the original game logic I wrote, using python-chess. The app no longer uses this.

import chess

class premoveChess:

    def __init__(self):
        self.board = chess.Board()
        self.gameState = chess.Board()
        validMove = False
        while(not validMove):
            try:
                self.writtenMove = self.getMove("White")
                self.gameState.push_san(self.writtenMove)
                validMove = True
            except:
                print("Not a valid move")
    
    def showBoard(self):#print board as players see it
        print(self.board)
    
    def showGameState(self):#print board as it will be after written move
        print(self.gameState)
    
    def getMove(self, turnPlayer):#This is where an interface will plug in(also the above two)
        return input(str(turnPlayer)+", enter your move")

    def startGame(self):#manage turn sequence
        result = "*"
        whiteToWriteMove = False
        while(result == "*"):
            result = self.turn(whiteToWriteMove)
            whiteToWriteMove = not whiteToWriteMove
        self.showGameState()
        print(result)
    
    def turn(self, whiteToWriteMove):#executes turn, returns game result
        self.showBoard()
        turnPlayer = "Black"
        if(whiteToWriteMove):
            turnPlayer = "White"
        try:
            newWrittenMove = self.getMove(turnPlayer)
            if(newWrittenMove == "draw"):
                return "Agreed draw"
            if("wins" in newWrittenMove):
                return newWrittenMove
            self.gameState.push_san(newWrittenMove)
        except:
            print("Illegal move!")
            if(whiteToWriteMove):
                return "0-1"
            else:
                return "1-0"
        self.board.push_san(self.writtenMove)
        self.writtenMove = newWrittenMove
        if(self.gameState.is_repetition() or self.gameState.halfmove_clock>99):
            return "Claimed draw"
        return self.gameState.result()

"""
Control flow is:
    White writes first move, then, until a written move ends the game:
        Black writes move
        White plays move
            First move is made freely like normal
            If black's written move is now illegal, white wins
            If black's written move ends the game, the move is played and the game ends
        White writes move
        Black plays move
            If white's written move is now illegal, black wins
            If white's written move ends the game, the move is played and the game ends
"""

File: premovechess.js
const { Chess } = require("chess.js");

class premoveChess{

    constructor(){
        this.board = new Chess();
        this.gameState = new Chess();
        this.writtenMove = {"from": "a1", "to": "b7"};
        this.toWriteMove = "White"
        this.status = "White to write initial move"
        this.gameOver = false;
    }

    showBoard(){
        return this.board.fen()
    }
    
    showGameState(){
        return this.gameState.fen();
    }

    playerToWriteMove(){
        return this.toWriteMove;
    }

    getStatus(){
        return this.status;
    }

    isGameOver(){
        return this.gameOver;
    }
    
    turn(move){//executes move, returns game status
        if(this.gameOver){
            return;
        }
        var valid = this.gameState.move(move);
        if(valid == null){
            if(this.toWriteMove=="White"){
                this.status =  "White tried to move "+move.from+" to "+move.to+", Black wins by illegal premove";
                this.gameOver=true;
                return;
            }
            else if(this.toWriteMove=="Black"){
                this.status =  "Black tried to move "+move.from+" to "+move.to+", White wins by illegal premove";
                this.gameOver=true;
                return;
            }
            else{
                this.status =  "Something went wrong here, India Papa";
                return;
            }
        }
        this.board.move(this.writtenMove);
        this.writtenMove = move;
        if(this.gameState.in_draw()){
            this.status =  this.toWriteMove+" moved "+move.from+" to "+move.to+", Draw by 50 moves or insufficient material";
            this.gameOver=true;
            return;
        }
        if(this.gameState.in_stalemate()){
            this.status =  this.toWriteMove+" moved "+move.from+" to "+move.to+", Draw by stalemate";
            this.gameOver=true;
            return;
        }
        if(this.gameState.in_threefold_repetition()){
            this.status =  this.toWriteMove+" moved "+move.from+" to "+move.to+", Draw by threefold repetition";
            this.gameOver=true;
            return;
        }
        if(this.gameState.in_checkmate()){
            this.status =  this.toWriteMove+" moved "+move.from+" to "+move.to+", "+this.toWriteMove+" wins by checkmate";
            this.gameOver=true;
            return;
        }
        if(this.toWriteMove=="White"){
            this.toWriteMove="Black";
        }
        else{
            this.toWriteMove="White";
        }
        this.status =  this.toWriteMove+" to write a move";
        return;
    }
}

module.exports = premoveChess;